IMPORTING LIBRARIES

In [1]:
import pandas as pd
import numpy as np
import os

IMPORTING DATA

In [5]:
# Creating Path
path = r'C:\Users\henry\OneDrive\042025Instacart Basket Analysis'

In [7]:
# Importing pkl file
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 DATA', 'Prepared Data', '.pkl'))


In [9]:
# Creating a subset of 1 million rows
df = ords_prods_merge[:1000000]

In [11]:
# Check shape of new subset
df.shape

(1000000, 17)

In [13]:
df.head(10)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   
5           1  Chocolate Sandwich Cookies        61             19     5.8   
6           1  Chocolate Sandwich Cookies        61             19     5.8   
7           1  Chocolate Sandwich Cookies        61             19     5.8   
8           1  Chocolate Sandwich Cookies        61             19     5.8   
9           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   3139998      138            28          6                 11   
1   1977647      138            30          6                 17   
2    389851      709             2          0                 21   
3    652770      764             1          3                 13   
4   1813452      764             3          4                 17   
5   1701441      777            16          1                  7   
6   1871483      825             3          2                 14   
7   1290456      910            12          3                 10   
8    369558     1052            10          1                 20   
9    589712     1052            15          1                 12   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     NaN                 10          0   both   
4                     9.0                 11          1   both   
5                    26.0                  7          0   both   
6                    30.0                  2          0   both   
7                    30.0                  1          0   both   
8                    19.0                  1          0   both   
9                    15.0                  2          1   both   

     price_range_loc     busiest_day busiest_period_of_day  
0  Mid-range product  Regularly busy           Most orders  
1  Mid-range product  Regularly busy        Average orders  
2  Mid-range product     Busiest day        Average orders  
3  Mid-range product  Regularly busy           Most orders  
4  Mid-range product      Least busy        Average orders  
5  Mid-range product  Regularly busy        Average orders  
6  Mid-range product  Regularly busy           Most orders  
7  Mid-range product  Regularly busy           Most orders  
8  Mid-range product  Regularly busy        Average orders  
9  Mid-range product  Regularly busy           Most orders

In [15]:
# Using the 'groupby()' function 
df.groupby('product_name')

In [17]:
# Aggregating Data with agg()
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.577493
2                17.320781
3                16.084944
4                17.530458
5                14.763075
6                16.658449
7                17.031590
8                15.076662
9                15.447580
10               18.681852
11               15.447411
12               14.327957
13               16.548642
14               16.960241
15               16.121948
16               17.803851
17               15.593633
18               19.674252
19               16.899756
20               16.255442
21               25.535479

When comparing the subset of 100,000 rows with the entire dataframe, a significant difference emerges in the coverage of departments. The full dataframe contains all 21 department_id. This broader coverage in the full dataset results in more comprehensive and representative trends. Additionally, the mean order_number for each department varies between the two datasets, reflecting different customer behaviors and purchase patterns.The subset may misrepresent certain customer ordering patterns. This highlights the importance of using the entire dataset for more accurate analysis.

In [20]:
# Aggregations without the use of the agg() function
df.groupby('department_id')['order_number'].mean()

department_id
1     15.577493
2     17.320781
3     16.084944
4     17.530458
5     14.763075
6     16.658449
7     17.031590
8     15.076662
9     15.447580
10    18.681852
11    15.447411
12    14.327957
13    16.548642
14    16.960241
15    16.121948
16    17.803851
17    15.593633
18    19.674252
19    16.899756
20    16.255442
21    25.535479
Name: order_number, dtype: float64

In [22]:
# Performing multiple aggregations
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                15.577493   1  99
2                17.320781   1  96
3                16.084944   1  99
4                17.530458   1  99
5                14.763075   1  99
6                16.658449   1  99
7                17.031590   1  99
8                15.076662   1  98
9                15.447580   1  99
10               18.681852   1  99
11               15.447411   1  99
12               14.327957   1  99
13               16.548642   1  99
14               16.960241   1  99
15               16.121948   1  99
16               17.803851   1  99
17               15.593633   1  99
18               19.674252   1  99
19               16.899756   1  99
20               16.255442   1  99
21               25.535479   1  98

Aggregating Data with transform function

In [25]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\henry\AppData\Local\Temp\ipykernel_23956\1681540670.py:1: FutureWarning: The provided callable <function max at 0x0000023977AF8400> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [29]:
ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  order_dow  order_hour_of_day  \
0    3139998      138            28          6                 11   
1    1977647      138            30          6                 17   
2     389851      709             2          0                 21   
3     652770      764             1          3                 13   
4    1813452      764             3          4                 17   
5    1701441      777            16          1                  7   
6    1871483      825             3          2                 14   
7    1290456      910            12          3                 10   
8     369558     1052            10          1                 20   
9     589712     1052            15          1                 12   
10   1798115     1379             1          5                 11   
11    497193     1480            36          3                 13   
12   1727826     1480            38          4                 13   
13   2694698     1480            39          2                 12   
14   3171753     1494             6          3                 10   

    days_since_prior_order  add_to_cart_order  reordered _merge  \
0                      3.0                  5          0   both   
1                     20.0                  1          1   both   
2                      6.0                 20          0   both   
3                      NaN                 10          0   both   
4                      9.0                 11          1   both   
5                     26.0                  7          0   both   
6                     30.0                  2          0   both   
7                     30.0                  1          0   both   
8                     19.0                  1          0   both   
9                     15.0                  2          1   both   
10                     NaN                  1          0   both   
11                     7.0                  1          0   both   
12                     7.0                  3          1   both   
13                    12.0                  4          1   both   
14                     2.0                  8          0   both   

      price_range_loc     busiest_day busiest_period_of_day  max_order  
0   Mid-range product  Regularly busy           Most orders         32  
1   Mid-range product  Regularly busy        Average orders         32  
2   Mid-range product     Busiest day        Average orders          5  
3   Mid-range product  Regularly busy           Most orders          3  
4   Mid-range product      Least busy        Average orders          3  
5   Mid-range product  Regularly busy        Average orders         26  
6   Mid-range product  Regularly busy           Mo

In [31]:
# first 100 rows
ords_prods_merge.head(100)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
..         ...                         ...       ...            ...     ...   
95           1  Chocolate Sandwich Cookies        61             19     5.8   
96           1  Chocolate Sandwich Cookies        61             19     5.8   
97           1  Chocolate Sandwich Cookies        61             19     5.8   
98           1  Chocolate Sandwich Cookies        61             19     5.8   
99           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  order_dow  order_hour_of_day  \
0    3139998      138            28          6                 11   
1    1977647      138            30          6                 17   
2     389851      709             2          0                 21   
3     652770      764             1          3                 13   
4    1813452      764             3          4                 17   
..       ...      ...           ...        ...                ...   
95    602103    10831             8          3                 11   
96     49629    11119            18          1                 14   
97    317888    11186            13          5                 16   
98    682486    11243            16          3                 13   
99   2861386    11448            10          2                 15   

    days_since_prior_order  add_to_cart_order  reordered _merge  \
0                      3.0                  5          0   both   
1                     20.0                  1          1   both   
2                      6.0                 20          0   both   
3                      NaN                 10          0   both   
4                      9.0                 11          1   both   
..                     ...                ...        ...    ...   
95                    23.0                  5          0   both   
96                    30.0                  1          0   both   
97                     2.0                  8          0   both   
98                     0.0                  2          0   both   
99                     3.0                  6          0   both   

      price_range_loc     busiest_day busiest_period_of_day  max_order  
0   Mid-range product  Regularly busy           Most orders         32  
1   Mid-range product  Regularly busy        Average orders         32  
2   Mid-range product     Busiest day        Average orders          5  
3   Mid-range product  Regularly busy           Most orders          3  
4   Mid-range product      Least busy        Average orders          3  
..                ...             ...                   ...        ...  
95  Mid-range product  Regularly busy           Most orders         10  
96  Mid-range product  Regularly busy           Most orders         23  
97  Mid-range product  Regularly busy           Most orders         26  
98  Mid-range product  Regularly busy           Most orders         43  
99  Mid-range product  Regularly busy           Most orders         21  

[100 rows x 18 columns]

Creating loyalty labels or flags

In [33]:

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [35]:
# Check output
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15874128
Loyal customer      10282763
New customer         6242841
Name: count, dtype: int64

In [37]:
# Check Loyalty flag in dataframe
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number','max_order']].head(15)

user_id      loyalty_flag  order_number  max_order
0       138  Regular customer            28         32
1       138  Regular customer            30         32
2       709      New customer             2          5
3       764      New customer             1          3
4       764      New customer             3          3
5       777  Regular customer            16         26
6       825      New customer             3          9
7       910  Regular customer            12         12
8      1052  Regular customer            10         20
9      1052  Regular customer            15         20
10     1379      New customer             1         10
11     1480    Loyal customer            36         43
12     1480    Loyal customer            38         43
13     1480    Loyal customer            39         43
14     1494  Regular customer             6         13

In [39]:
# Basic statistics of each loyalty flag
ords_prods_merge.groupby(['loyalty_flag']).agg({'prices':['mean','min','max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773575  1.0  25.0
New customer      7.801206  1.0  25.0
Regular customer  7.798262  1.0  25.0

The average prices are nearly identical across loyalty groups: Loyal customers spend $7.77, new customers $7.80, and regular customers $7.80. Minimum and maximum prices are consistent at $1.00 and $25.00. This suggests loyalty status does not significantly influence product-level spending.

Column for average spender

In [42]:
# Creating column for average spender
ords_prods_merge['average_spend']=ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\henry\AppData\Local\Temp\ipykernel_23956\655122787.py:2: FutureWarning: The provided callable <function mean at 0x0000023977AF8E00> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['average_spend']=ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [44]:
# Creating flag for low and high spender
ords_prods_merge.loc[ords_prods_merge['average_spend']<10,'spender_flag']='Low_spender'

ords_prods_merge.loc[ords_prods_merge['average_spend']>= 10, 'spender_flag'] = 'High_spender'

In [46]:
# Check output
ords_prods_merge['spender_flag'].value_counts(dropna = False)

spender_flag
Low_spender     32280013
High_spender      119719
Name: count, dtype: int64

In [52]:
ords_prods_merge[['user_id', 'average_spend', 'spender_flag', 'prices']].head(50)

user_id  average_spend spender_flag  prices
0       138       6.935811  Low_spender     5.8
1       138       6.935811  Low_spender     5.8
2       709       7.930208  Low_spender     5.8
3       764       4.972414  Low_spender     5.8
4       764       4.972414  Low_spender     5.8
5       777       6.935398  Low_spender     5.8
6       825       5.957576  Low_spender     5.8
7       910       6.680000  Low_spender     5.8
8      1052       7.162500  Low_spender     5.8
9      1052       7.162500  Low_spender     5.8
10     1379       5.813636  Low_spender     5.8
11     1480       5.081633  Low_spender     5.8
12     1480       5.081633  Low_spender     5.8
13     1480       5.081633  Low_spender     5.8
14     1494       6.803797  Low_spender     5.8
15     1494       6.803797  Low_spender     5.8
16     1494       6.803797  Low_spender     5.8
17     1540       6.607122  Low_spender     5.8
18     1540       6.607122  Low_spender     5.8
19     1540       6.607122  Low_spender     5.8
20     1540       6.607122  Low_spender     5.8
21     1540       6.607122  Low_spender     5.8
22     1540       6.607122  Low_spender     5.8
23     1540       6.607122  Low_spender     5.8
24     1540       6.607122  Low_spender     5.8
25     1540       6.607122  Low_spender     5.8
26     1540       6.607122  Low_spender     5.8
27     1540       6.607122  Low_spender     5.8
28     1540       6.607122  Low_spender     5.8
29     1540       6.607122  Low_spender     5.8
30     1540       6.607122  Low_spender     5.8
31     1540       6.607122  Low_spender     5.8
32     1540       6.607122  Low_spender     5.8
33     1540       6.607122  Low_spender     5.8
34     1598       7.123529  Low_spender     5.8
35     1598       7.123529  Low_spender     5.8
36     1598       7.123529  Low_spender     5.8
37     1647       4.853521  Low_spender     5.8
38     1647       4.853521  Low_spender     5.8
39     1647       4.853521  Low_spender     5.8
40     1647       4.853521  Low_spender     5.8
41     1647       4.853521  Low_spender     5.8
42     2833       7.377143  Low_spender     5.8
43     2850       8.054153  Low_spender     5.8
44     2857       8.353793  Low_spender     5.8
45     3010       6.452800  Low_spender     5.8
46     3029       7.960606  Low_spender     5.8
47     3395       6.322222  Low_spender     5.8
48     3904       7.511111  Low_spender     5.8
49     4006       7.441100  Low_spender     5.8

Columns for frequent users

In [54]:
# Creating a column for frequent users
ords_prods_merge['Customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\henry\AppData\Local\Temp\ipykernel_23956\928642412.py:2: FutureWarning: The provided callable <function median at 0x0000023977D0FA60> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['Customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [56]:
# Creating frequency flags
ords_prods_merge.loc[ords_prods_merge['Customer_frequency'] > 20, 'frequency_flag'] = 'Non_frequent_customer'

ords_prods_merge.loc[(ords_prods_merge['Customer_frequency']>10) & (ords_prods_merge['Customer_frequency']<=20),'frequency_flag'] = 'Regular customer'

ords_prods_merge.loc[ords_prods_merge['Customer_frequency'] <= 10, 'frequency_flag'] = 'Frequent Customer'

In [58]:
# check output
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent Customer        21556644
Regular customer          7207340
Non_frequent_customer     3635743
NaN                             5
Name: count, dtype: int64

In [60]:
ords_prods_merge[['user_id','days_since_prior_order', 'Customer_frequency', 'frequency_flag']].head(50)

user_id  days_since_prior_order  Customer_frequency         frequency_flag
0       138                     3.0                 8.0      Frequent Customer
1       138                    20.0                 8.0      Frequent Customer
2       709                     6.0                 8.0      Frequent Customer
3       764                     NaN                 9.0      Frequent Customer
4       764                     9.0                 9.0      Frequent Customer
5       777                    26.0                11.0       Regular customer
6       825                    30.0                20.0       Regular customer
7       910                    30.0                 6.0      Frequent Customer
8      1052                    19.0                10.0      Frequent Customer
9      1052                    15.0                10.0      Frequent Customer
10     1379                     NaN                20.0       Regular customer
11     1480                     7.0                 7.0      Frequent Customer
12     1480                     7.0                 7.0      Frequent Customer
13     1480                    12.0                 7.0      Frequent Customer
14     1494                     2.0                 7.0      Frequent Customer
15     1494                     6.0                 7.0      Frequent Customer
16     1494                    10.0                 7.0      Frequent Customer
17     1540                     7.0                 7.0      Frequent Customer
18     1540                     6.0                 7.0      Frequent Customer
19     1540                    14.0                 7.0      Frequent Customer
20     1540                     8.0                 7.0      Frequent Customer
21     1540                     6.0                 7.0      Frequent Customer
22     1540                    28.0                 7.0      Frequent Customer
23     1540                     7.0                 7.0      Frequent Customer
24     1540                     7.0                 7.0      Frequent Customer
25     1540                     0.0                 7.0      Frequent Customer
26     1540                     7.0                 7.0      Frequent Customer
27     1540                     7.0                 7.0      Frequent Customer
28     1540                     7.0                 7.0      Frequent Customer
29     1540                     7.0                 7.0      Frequent Customer
30     1540                     3.0                 7.0      Frequent Customer
31     1540                    16.0                 7.0      Frequent Customer
32     1540                     5.0                 7.0      Frequent Customer
33     1540                     6.0                 7.0      Frequent Customer
34     1598                     NaN                 2.0      Frequent Customer
35     1598                    21.0                 2.0      Frequent Customer
36     1598                     2.0                 2.0      Frequent Customer
37     1647                     NaN                25.0  Non_frequent_customer
38     1647                    18.0                25.0  Non_frequent_customer
39     1647                     8.0                25.0  Non_frequent_customer
40     1647                    30.0                25.0  Non_frequent_customer
41     1647                    25.0                25.0  Non_frequent_customer
42     2833                     NaN                13.0       Regular customer
43     2850                     9.0                 7.0      Frequent Customer
44     2857                    10.0                11.0       Regular customer
45     3010                     7.0                 7.0      Frequent Customer
46     3029                     NaN                28.0  Non_frequent_customer
47     3395                    30.0                18.0       Regular customer
48     3904                     NaN                21.0  Non_frequent_customer
49     4006                     7.0                 7.

EXPORT DATA

In [63]:
ords_prods_merge.to_pickle(os.path.join(path, '02 DATA', 'Prepared Data', 'Aggregated_ords_prods_merge.pkl'))